File: erettsegi_emelt/2020_okt/f#/Sorozatok.fs
open System
open System.IO
open System.Globalization

let napok = [| "v"; "h"; "k"; "sz"; "cs"; "p"; "szo" |]
let honapok = [| 0; 3; 2; 5; 0; 3; 5; 1; 4; 6; 2; 4 |]

let hetnapja(ev: int, ho: int, nap: int) =
    let ev2 = if ho < 3 then ev - 1 else ev

    napok.[(ev2 + ev2 / 4 - ev2 / 100 + ev2 / 400 + honapok.[ho - 1] + nap) % 7]

let sorozatotKeszit(lines: string[], index: int) =
    let datumStr = lines.[index]
    let epizodInfoSplit = lines.[index + 2].Split('x')

    {|
        AdasbaKerulesiDatum = if datumStr = "NI" then DateTime.MinValue else DateTime.ParseExact(datumStr, "yyyy.MM.dd", CultureInfo.InvariantCulture);
        Cim = lines.[index + 1];
        EvadokSzama = int epizodInfoSplit.[0];
        EpizodokSzama = int epizodInfoSplit.[1];
        EpizodonkentiHossz = int lines.[index + 3]
        LattaEMarAKeszito = int lines.[index + 4] = 1
    |}

[<EntryPoint>]
let main _ =
    let lines = File.ReadAllLines("lista.txt")
    let sorozatok = seq { 0 .. 5 .. lines.Length - 1 } |> Seq.map(fun i -> sorozatotKeszit(lines, i))
                                                       |> Seq.toArray

    sorozatok |> Seq.filter(fun k -> k.AdasbaKerulesiDatum <> DateTime.MinValue)
              |> Seq.length
              |> printfn "2. Feladat: %d db ismert dátumú epizód van"

    sorozatok |> Seq.filter(fun k -> k.LattaEMarAKeszito)
              |> Seq.length
              |> fun k -> float k / float sorozatok.Length * 100.0
              |> printfn "3. Feladat: Látottak százaléka: %.2f%%"

    sorozatok |> Seq.filter(fun k -> k.LattaEMarAKeszito)
              |> Seq.sumBy(fun k -> k.EpizodonkentiHossz)
              |> float
              |> TimeSpan.FromMinutes
              |> fun k -> printfn "4. Feladat: Eltöltött idő: %d nap, %d óra és %d perc" k.Days k.Hours k.Minutes

    printfn "5. Feladat: Írj be 1 dátumot! (éééé.hh.nn)"

    let bekertDatumStr = Console.ReadLine()
    let bekertDatum = DateTime.ParseExact(bekertDatumStr, "yyyy.MM.dd", CultureInfo.InvariantCulture)

    sorozatok |> Seq.filter(fun k -> k.AdasbaKerulesiDatum <> DateTime.MinValue && not k.LattaEMarAKeszito)
              |> Seq.filter(fun k -> k.AdasbaKerulesiDatum < bekertDatum || k.AdasbaKerulesiDatum = bekertDatum)
              |> Seq.iter(fun k -> printfn "%dx%d\t%s" k.EvadokSzama k.EpizodokSzama k.Cim)

    printfn "7. Feladat: Add meg 1 hét napját! (h, k, sze, cs, p, szo, v)"

    let bekertNap = Console.ReadLine()

    sorozatok |> Seq.filter(fun k -> k.AdasbaKerulesiDatum <> DateTime.MinValue)
              |> Seq.filter(fun k -> bekertNap = hetnapja(k.AdasbaKerulesiDatum.Year, k.AdasbaKerulesiDatum.Month, k.AdasbaKerulesiDatum.Day))
              |> Seq.map(fun k -> k.Cim)
              |> Seq.distinct
              |> Seq.toArray
              |> fun k -> if k.Length = 0 then "Az adott napon nem kerül adásba sorozat." else String.Join("\n", k)
              |> printfn "%s"

    sorozatok |> Seq.groupBy(fun k -> k.Cim)
              |> Seq.map(fun (k, v) -> k + " " + string (v |> Seq.sumBy(fun m -> m.EpizodonkentiHossz)) + " " + string (v |> Seq.length))
              |> Seq.toArray
              |> fun k -> File.WriteAllLines("summa.txt", k)
    0

File: erettsegi_emelt/2020_okt/python/Sorozatok.py
from datetime import date, time, timedelta
from itertools import groupby

class Sorozat:
    def __init__(self, lines, index):
        datumStr = lines[index].strip()
        epizodInfoSplit = lines[index + 2].split('x')

        self.adasbaKerulesiDatum = date.min if datumStr == 'NI' else date.fromisoformat(datumStr.replace('.', '-'))
        self.cim = lines[index + 1].strip()
        self.evadokSzama = int(epizodInfoSplit[0])
        self.epizodokSzama = int(epizodInfoSplit[1])
        self.epizodonkentiHossz = int(lines[index + 3])
        self.lattaEMarAKeszito = int(lines[index + 4]) == 1

napok = [ "v", "h", "k", "sz", "cs", "p", "szo" ]
honapok = [ 0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4 ]

def hetnapja(ev, ho, nap):
    ev = ev - 1 if ho < 3 else ev

    return napok[(ev + ev // 4 - ev // 100 + ev // 400 + honapok[ho - 1] + nap) % 7]


with open('lista.txt', 'r') as file:
    lines = file.readlines()
    sorozatok = [ Sorozat(lines, i) for i in range(0, len(lines), 5) ]

ismertDatumuakSzama = sum(1 for k in sorozatok if k.adasbaKerulesiDatum != date.min)
keszitoAltalLatottakSzama = sum(1 for k in sorozatok if k.lattaEMarAKeszito)

print(f'2. Feladat: {ismertDatumuakSzama} db ismert dátumú epizód van')
print('3. Feladat: Látottak százaléka: %.2f%%' % (keszitoAltalLatottakSzama / len(sorozatok) * 100))

osszesElpazaroltPerc = sum(k.epizodonkentiHossz for k in sorozatok if k.lattaEMarAKeszito)
elpazaroltIdo = timedelta(minutes = osszesElpazaroltPerc)
orak, maradek = divmod(elpazaroltIdo.seconds, 3600)
percek, _ = divmod(maradek, 60)

print(f'4. Feladat: Eltöltött idő: {elpazaroltIdo.days} nap, {orak} óra és {percek} perc')

bekertDatumStr = input('5. Feladat: Írj be 1 dátumot! (éééé.hh.nn) ')
bekertDatum = date.fromisoformat(bekertDatumStr.replace('.', '-'))

for sorozat in sorozatok:
    if sorozat.adasbaKerulesiDatum != date.min and sorozat.adasbaKerulesiDatum <= bekertDatum and not sorozat.lattaEMarAKeszito:
        print(f'{sorozat.evadokSzama}x{sorozat.epizodokSzama}\t{sorozat.cim}')

bekertNap = input('7. Feladat: Add meg 1 hét napját! (h, k, sze, cs, p, szo, v)' )
bekertNapraEsok = set(k.cim for k in sorozatok if k.adasbaKerulesiDatum != date.min and bekertNap == hetnapja(k.adasbaKerulesiDatum.year, k.adasbaKerulesiDatum.month, k.adasbaKerulesiDatum.day))

print('Az adott napon nem kerül adásba sorozat.' if len(bekertNapraEsok) == 0 else '\n'.join(bekertNapraEsok))

cimenkentiStatolo = lambda sorozatokCimhez: ( sum(k.epizodonkentiHossz for k in sorozatokCimhez), len(sorozatokCimhez) )
stat = dict((cim, cimenkentiStatolo(list(sorozatokCimhez))) for cim, sorozatokCimhez in groupby(sorozatok, key = lambda k: k.cim))

with open('summa.txt', 'w') as output:
    for cim, stat in stat.items():
        output.write(f'{cim} {stat[0]} {stat[1]}\n')